### ISSUE TITLE
Add the ability to use internal controllers in modules

### ISSUE THREAD

**Issue by ChristopherHaws** [OrchardCMS/OrchardCore#11426]
### Is your feature request related to a problem? Please describe.
In C#, to prevent other developers from using types in your modules/dlls that they shouldn't be using, you should mark them as `internal`. Unfortunately, IoC containers such as `Microsoft.Extensions.DependencyInjection` do not work with internal constructors (for various reasons, the main one being that usually you expose internal constructors for unit testing and don't want the IoC container to think it can use them). Since we are limited to public constructors, the only way to inject a service via IoC is to either make the injected service public or make the service being resolved (in this case a controller) as internal. Since we don't want to expose every type in a module as public, the only option is to make the controller internal.

In standard ASP.NET Core MVC this is possible via:
```csharp
builder.Services.AddMvc().ConfigureApplicationPartManager(mvc => {
   // Add feature provider to allow "internal" controller
   mvc.FeatureProviders.Add(new InternalControllerFeatureProvider());
});

internal class InternalControllerFeatureProvider : ControllerFeatureProvider {
    protected override bool IsController(TypeInfo typeInfo) {
        var isCustomController = !typeInfo.IsAbstract
            && typeof(ControllerBase).IsAssignableFrom(typeInfo)
            && IsInternal(typeInfo);
        return isCustomController || base.IsController(typeInfo);

        bool IsInternal(Type t) =>
            !t.IsVisible
            && !t.IsPublic
            && t.IsNotPublic
            && !t.IsNested
            && !t.IsNestedPublic
            && !t.IsNestedFamily
            && !t.IsNestedPrivate
            && !t.IsNestedAssembly
            && !t.IsNestedFamORAssem
            && !t.IsNestedFamANDAssem;
    }
}
```

However in Orchard Core, [`.AddMvc` is called by Orchard](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Mvc.Core/Startup.cs#L138) and `IsController` is [pre-filtered by Orchard](https://github.com/OrchardCMS/OrchardCore/blob/743ecc6a0d186803e8b152e5585acee1a7442223/src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs#L285) to [only include public types](https://github.com/OrchardCMS/OrchardCore/blob/743ecc6a0d186803e8b152e5585acee1a7442223/src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs#L343).

See the following discussion for more info: https://github.com/OrchardCMS/OrchardCore/discussions/11417

### Describe the solution you'd like
I would like for there be a way to include internal types to the modules that contain the types. This could be enabled via something like:
```csharp
services.AddOrchardCore().AllowInternalTypes();
```
or
```csharp
services.AddOrchardCore(x => {
    x.AllowInternalTypes = true;
});
```

### Describe alternatives you've considered
The only alternative I have is to make all my internal types public.

Right now I am working around this using service locator patterns which I don't like:
```csharp
internal class UserDb : DbContext { /* ... */ }

public class UserController {
    private readonly UserDb db;

    public UserController(IServiceProvicer services) {
        this.db = services.GetRequiredService<UserDb>();
    }
}
```

**Comment by ns8482e** [comment]
@ChristopherHaws I am curious why you have to define `FeatureProviders` explicitly for internal controller?  

Is ASP.NET not considering  internal controller as it does for public controller?

**Comment by ChristopherHaws** [comment]
@ns8482e Correct, ASP.NET Core does not route to internal controllers by default.
